src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java

```diff
@@ -235,10 +234,11 @@ public QuerySearchResultProvider load(final ShardSearchRequest request, final Se
             }
         } else {
             key.shard.queryCache().onHit();
+            // restore the cached query result into the context
+            final QuerySearchResult result = context.queryResult();
+            result.readFromWithId(context.id(), value.streamInput());
```

Comment:
I think the only thing we need to know is whether it's safe to free the context immediately. Maybe we can add a flag on the context for that and not have to deserialize everything?